client/src/main/scala/org/http4s/client/PoolManager.scala

```diff
@@ -20,13 +22,33 @@ private final class PoolManager[F[_], A <: Connection[F]](
   private[this] val logger = getLogger(classOf[PoolManager[F, A]])
 
   private var isClosed = false
-  private var allocated = 0
-  private val idleQueue = new mutable.Queue[A]
-  private val waitQueue = new mutable.Queue[Waiting]
+  private var curTotal = 0
+  private var allocated = mutable.Map.empty[RequestKey, Int]
+  private val idleQueue = mutable.Map.empty[RequestKey, mutable.Queue[A]]
+  private val waitQueue = mutable.Queue.empty[Waiting]
 
   private def stats =
     s"allocated=$allocated idleQueue.size=${idleQueue.size} waitQueue.size=${waitQueue.size}"
 
+  private def getConnectionFromQueue(key: RequestKey): Option[A] =
+    idleQueue.get(key).flatMap(q => if (q.nonEmpty) Some(q.dequeue()) else None)
```

Comment:
We should also remove the `Queue` from `idleQueue` if it's empty or else we have a memory leak if folks are using lots of different hosts.

Reply:
We do that when the last connection to a host is dropped.

Reply:
As is, we never drop keys from `idleQueue`, only `allocated`. It might be a good idea to rename `idleQueue` to `idleQueues` since there are now more than one.